pkg/daemon/ceph/multus/mover.go

```diff
@@ -0,0 +1,437 @@
+		Namespace:          m.Namespace, // limit watch to namespace
+		MetricsBindAddress: "0",         // disable metrics so ports don't collide on host
+	})
+	if err != nil {
+		return errors.Wrapf(err, "failed to create manager for %q controller", MoverControllerName)
+	}
+
+	movedInterfaceCache := NewMovedInterfaceCache()
+	var networkMutex sync.Mutex
+
+	reconciler := &HolderPodReconciler{
+		Client:       mgr.GetClient(),
+		NetworkMutex: &networkMutex,
+		MovedCache:   movedInterfaceCache,
+	}
+
+	err = controllerruntime.
+		NewControllerManagedBy(mgr).
+		Named(MoverControllerName).
+		For(&corev1.Pod{}). // limit watch to pods
+		WithEventFilter(predicateAcceptOnlyHolderPods(m.NodeName)).
+		WithOptions(controller.Options{
+			// retry at least every 10 seconds so we don't make CSI wait longer than that in the
+			// worst case, but don't bother retrying more often than once every 1/2 second
+			RateLimiter: workqueue.NewItemExponentialFailureRateLimiter(500*time.Millisecond, 10*time.Second),
+		}).
+		Complete(reconciler)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create %q controller", MoverControllerName)
+	}
+
+	ctx, cancelCtx := context.WithCancel(ctx)
+	defer cancelCtx()
+
+	// perform an initial best-effort clean-up in case a previous mover instance died without cleaning up
+	cleanupAllInterfacesWithPrefix(MovedNetworkInterfacePrefix)
+
+	// before starting the controller, register a routine to clean up when we return
+	// importantly, this will clean up even if the routine panics
+	defer func() {
+		logger.Info("removing multus interfaces from host network namespace before terminating mover process")
+
+		// Stop any ongoing reconciles (important if this is a cleanup due to a panic)
+		cancelCtx()
+
+		// treat the full cleanup as an atomic action that should block the reconciler from
+		// creating more interface copies during teardown. If we get this lock, we can be sure there
+		// is no current reconcile copying interfaces, and because the context is canceled we
+		// can be sure there will be no subsequent reconciles after we unlock.
+		networkMutex.Lock()
+		defer networkMutex.Unlock()
+
+		for holderPodName, info := range movedInterfaceCache.AsMap() {
+			logger.Infof("un-moving interface for holder pod %q with info: %+v", holderPodName, info)
+
+			if err := cleanUpMovedInterface(holderPodName, movedInterfaceCache); err != nil {
+				logger.Errorf("failed to clean up interface for holder pod %q during termination; "+
+					"continuing despite failure", holderPodName)
+			}
+		}
+
+		// clean up any remaining interfaces that might not have been cleaned up for some reason
+		cleanupAllInterfacesWithPrefix(MovedNetworkInterfacePrefix)
+	}()
+
+	// NOTE: this blocks until the context is canceled
+	err = mgr.Start(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "failed to start %q controller", MoverControllerName)
+	}
+
+	return nil
+}
+
+type HolderPodReconciler struct {
+	client.Client
```

Comment:
Do we need to add the controller context?

Reply:
It's not necessary. Controller-runtime has the context and passes it to the `Reconcile()` method below.